### PR TITLE
YM-479 | Use HDS styled email templates

### PR DESCRIPTION
**Pull request by Jokinen**
### Context

Refactors the HTML templates in the application to use HDS styles.

HTML templates in email come with some caveats. Mainly, the markup that is supported varies a lot. To support most of the largest email programs, the markup that can be used is ancient. In essence we are forced to use inline styles which causes a lot of duplicate definitions.

To make the maintenance at least somewhat bearable, we are adding some build tooling for the email templates. In this context we are using Jinja templates which allows the email templates to be defined in a more DRY way. This should also make it easier to maintain the templates as the actual content is not hidden within a thousand lines of HTML tables.

The pretty meaningful downside is that this makes it relatively impossible for non-technical people to update these templates. Ideally, perhaps, this would be possible. It's possible to build more tooling on top of this system to allow that, but it would require a lot of extra work. This is definitely a system that's not ideal if we would want Jässäri people to maintain the templates.

I've tested the template with some email applications. I noticed that at least some styles are broken--but this is to be expected in my opinion. I'm willing to fix any style error which are too "crass". You can inspect the designs here:

* https://app.abstract.com/projects/b80d4d50-787f-11e9-ae1d-ef140a99b44b/branches/master/commits/603f5b5680a6721a76d2c84bce7b1c317504b1dc/files/651BDE24-1AC6-4C50-9D13-64A3D865F22B/layers/28E501BB-8ECC-4D6F-9469-C01FE84F14A2?mode=build
* https://share.goabstract.com/8a99555d-b09b-4273-9c09-9bfbea7a7c70

### Technical details

* Creates email templates in the `templates` folder
* Adds logic for rendering notifications from jinja templates
* Adds a management command that re-creates notifications on container start
* Changes some notification contents to be configurable from settings
    * Because notifications should no longer be edited by using the admin, we need to offer a build/run time mechanism for creating environment specific flavours of notifications. GHA has been configured to provide correct configs for staging and production.
* Removes notification seeding logic 
* Adds a management command (`send_test_mail`) that allows email sending to be tested
    * This command can be used to attain the HTML that is generated based on the notification template. You can then use the markup to test the templating behaviour, and the compatibility with various email programs.

The review build should have the built templates visible in the admin:
https://youth-membership-feature-ym-479-56.test.kuva.hel.ninja/admin/django_ilmoitin/notificationtemplate/1/change/

### CODE REVIEW

File: common_utils/utils.py [hunk 1]
```diff
@@ -1,8 +1,34 @@
...
+from youths.enums import NotificationType as YouthNotificationType
+
+logger = logging.getLogger(__name__)
+handler = logging.StreamHandler()
+logger.setLevel(logging.DEBUG)
+# create console handler and set level to debug
+handler.setLevel(logging.DEBUG)
+
+# add ch to logger
+logger.addHandler(handler)
```
Comment (Jokinen) [inline common_utils/utils.py:24]: Is there a better way to configure logging? This doesn't look correct to me.

File: common_utils/utils.py [hunk 2]
```diff
@@ -31,3 +57,147 @@ def get_original_client_ip():
...
+
+        try:
+            rendered = template.render(
+                image_location=settings.EMAIL_TEMPLATE_IMAGE_SOURCE
+            )
+
+            return rendered
+
+        except Exception:
+            print("Unexpected error:", sys.exc_info()[0])
```
Comment (Jokinen) [inline common_utils/utils.py:77]: is this good enough in this context?

File: common_utils/tests/test_commands.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from django.core.management import call_command
+from django_ilmoitin.models import NotificationTemplate
+
+
+def test_command_generate_notifications_from_templates_generates_notifications():
+    assert NotificationTemplate.objects.count() == 0
+
+    call_command("generate_notifications_from_templates")
+
+    assert NotificationTemplate.objects.count() > 0
```
Comment (Jokinen) [inline common_utils/tests/test_commands.py:10]: Any other cases I should test?

File: templates/email/messages/youth_profile_confirmed_en.html [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+{% raw %}
+Welcome to the Youth Services of the City of Helsinki! {{ youth_profile.approver_first_name }} has approved your membership. Your digital memberhsip card can be found here:
+</br>
+</br>
+{% endraw %}
+{{ macros.buttonLink(label="Jässäri", href="{{ youth_membership_ui_base_url }}/en/") }}
+{% raw %}
+</br>
+</br>
+For more information on the Youth Services membership and its perks, please visit https://jassari.munstadi.fi/en/.
```
Comment (Jokinen) [inline templates/email/messages/youth_profile_confirmed_en.html:13]: This link should be wrapped in an anchor.

File: common_utils/tests/test_commands.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    call_command("generate_notifications_from_templates", "--dev")
+
+    a_template_file = None
+
+    with open(
+        os.path.join(EMAIL_GENERATED_PATH, "youth_profile_confirmed_en.html"), "r"
+    ) as generated_template:
+        a_template_file = generated_template.read()
+
+    snapshot.assert_match(a_template_file)
```
Comment (Jokinen) [inline common_utils/tests/test_commands.py:31]: Is there a better approach than a snapshot available here?

File: common_utils/utils.py (hunk 1 above)
Comment (charn) [reply common_utils/utils.py:24]: There is a better way. You should do the in the settings.py (there's the LOGGING setting).

In this particular case I would add the following 

```python
LOGGING = {
    ...
    "loggers": {
        ...
        "common_utils.utils": {
            "handlers": ["console"],
            "level": "INFO",
        },
    },
}
```

File: common_utils/utils.py (hunk 2 above)
Comment (charn) [reply common_utils/utils.py:77]: Depending on how the logging is configured, the following might automagically also print the exception.
```python
except Exception:
    logger.exception("Error in generating the templates.")
```

File: common_utils/management/commands/send_test_mail.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            type=json.loads,
+            default={},
+        )
+
+    def handle(self, *args, **options):
+        email = options["email"]
+        notification_type = options["notification_type"]
+        custom_context = options["context"]
+
+        if notification_data := get_notification_data(notification_type):
```
Comment (charn) [inline common_utils/management/commands/send_test_mail.py:63]: Haven't seen these that much yet, but then again they just appeared in Python 3.8. :+1:

File: docker-entrypoint.sh [hunk 7]
```diff
@@ -22,6 +22,8 @@ if [[ "$CREATE_ADMIN_USER" = "1" ]]; then
     fi
 fi
 
+python /app/manage.py generate_notifications_from_templates
```
Comment (charn) [inline docker-entrypoint.sh:25]: I think the problem with this being here would be that if more than one container is run multiple times (which happens for staging and production). Would moving this to `on_deploy.sh` work?

File: common_utils/tests/test_commands.py (hunk 3 above)
Comment (charn) [reply common_utils/tests/test_commands.py:15]: If you want I guess you could make it a tiny bit more explicit by testing that correct amount of templates and translations are generated (similarly how it was in the previous tests)

```
    #from django.apps import apps
    NotificationTemplateTranslation = apps.get_model(  # noqa: N806
        "django_ilmoitin", "NotificationTemplateTranslation"
    )

    assert NotificationTemplate.objects.count() == 2
    assert NotificationTemplateTranslation.objects.count() == 6
```

File: common_utils/tests/test_commands.py (hunk 5 above)
Comment (charn) [reply common_utils/tests/test_commands.py:34]: Snapshot might have the side-effect that you have to update the generated snapshot data for every change. You could instead test some relevant information is in the template like:
```python
assert "<title>Welcome to the Youth Services of the City of Helsinki!</title>" in a_template_file
```

But I'm not sure if that's sufficient.

**Comment by charn** [comment]
As mentioned IRL I pushed a refactoring commit with the following changes:

- Use Jinja2 by configuring it as a template backend
- Move email templates so that Jinja2 backend is automatically used
- `docker-entrypoint.sh` runs `seed_data` and `generate_notifications_from_templates`
  only when migrations are being run (which is happens usually on development environments, production uses `on_deploy.sh`)
- `on_deploy.sh` runs `generate_notifications_from_templates`

**Review by charn — APPROVED** [review]
As far as I know, this looks good to me! :+1: :shipit:

File: templates/jinja2/email/macros.jinja [hunk 8]
```diff
@@ -17,7 +17,7 @@
 {%- endmacro %}
 
 {% macro logo(variant="header", lang="fi") -%}
-{% with image_path, alt = [image_location + "/helsinki-logo-" + variant + "-" + lang + ".png", "Helsinki" if lang == "fi" else "Helsingfors"] %}
+{% with image_path, alt = [image_location + "/helsinki-logo-" + variant + "-" + lang + ".png", "Helsingfors" if lang == "sv" else "Helsinki"] %}
```
Comment (Jokinen) [inline templates/jinja2/email/macros.jinja:20]: Good catch 👍

**Comment by Jokinen** [comment]
The approach is much cleaner now 👍

**Pull request merged**